protocols/noise/src/io.rs

```diff
@@ -121,57 +121,74 @@ enum WriteState {
     EncErr
 }
 
-impl<T: io::Read> io::Read for NoiseOutput<T> {
-    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        let buffer = self.buffer.borrow_mut();
+impl<T: AsyncRead + Unpin> AsyncRead for NoiseOutput<T> {
+    fn poll_read(mut self: core::pin::Pin<&mut Self>, cx: &mut futures::task::Context<'_>, buf: &mut [u8]) -> Poll<Result<usize, futures::io::Error>> {
+        // We use a `this` because the compiler isn't smart enough to allow
+        // mutably borrowing multiple different fields from the `Pin` at the
+        // same time.
```

Comment:
I don't think there's documentation about that.

The problem here is that `self` is a `Pin`, and `Pin<T>` implements `DerefMut<Target = T>`.
But when you do `self.foo` (to access the field `foo`), the compiler calls `deref_mut` on `self`, which entirely borrows `self`. If you try to access a different field, then it does the same and complains that `self` is already borrowed.

The fact that the compiler allows simultaneously borrowing multiple different fields from the same object is kind of a hack provided by the compiler, and doesn't work across the `DerefMut` trait.

Reply:
> The problem here is that self is a Pin, and Pin<T> implements DerefMut<Target = T>.

I see, I'm familiar with that kind of problem from calling other `&mut self` methods from inside a `&mut self` method even though both mutably borrow completely distinct fields of `Self`. I guess it was just wishful thinking on my part that de-referencing a `Pin` (or any type with `Deref`) on `Self` would be treated specially, i.e. could be elided like a newtype and not require going through regular deref-coercions.

> But when you do self.foo (to access the field foo), the compiler calls deref_mut [..]

To my understanding, insertion of `deref()` / `deref_mut()` calls (i.e. deref-coercions) are only inserted by the compiler if they resolve a type mismatch, so if `self: &mut Self` and `Self` indeed has the field `foo`, then `self.foo` does not involve `std::ops::Deref(Mut)`. On the other hand, if `self: Pin<&Self>` and it doesn't get special handling, it makes sense, of course, that `self.foo` is desugared to `(*Deref::deref(&self)).foo` and thus borrows `self` in its entirety. From that point of view, I probably wouldn't call the ability to simultaneously borrow multiple fields a hack though.

Reply:
So, I would suggest to remove these "the compiler isn't smart enough" comments, which I find slightly irritating, because this isn't really anything new, it's just the well-known difference between

```rust
let x = &mut self.deref_mut().x;
let y = &mut self.deref_mut().y; // error
```
and
```rust
let mut this = self.deref_mut();
let x = &mut this.x;
let y = &mut this.y; // ok
```

Explicitly writing `let mut this = self.deref_mut();` should make this clear enough.

Reply:
I am fine with either way. I just think we need to be consistent across the codebase.

Reply:
> I am fine with either way. I just think we need to be consistent across the codebase.

Which two competing approaches do you see here? I'm just saying that I think I would not have stumbled over this part of the code were it not for the presence of the (in my opinion misleading) code comment, which made me think for a moment that there is a compiler / borrow-checker limitation in place here that is somehow specific to the usage of `Pin` and deserves special mention, when in fact there is nothing unusual at all going on.